mqtt/mqtt-bridge/tests/integration.rs

```diff
@@ -0,0 +1,418 @@
+
+    // delay to have authorizer updated
+    tokio::time::delay_for(Duration::from_secs(2)).await;
+
+    // Send upstream
+    upstream_client
+        .publish_qos1("to/filter/1", "from upstream after update", false)
+        .await;
+
+    assert_matches!(
+        local_client.publications().recv().await,
+        Some(ReceivedPublication{payload, .. }) if payload == Bytes::from("from upstream after update")
+    );
+}
+
+#[tokio::test]
+async fn connect_to_upstream_failure_should_retry() {
+    let (local_server_handle, _) = setup_local_broker(AllowAll);
+
+    let subs = vec![
+        Direction::Out(TopicRule::new(
+            "temp/#".into(),
+            Some("to".into()),
+            Some("upstream".into()),
+        )),
+        Direction::In(TopicRule::new(
+            "filter/#".into(),
+            Some("to".into()),
+            Some("downstream".into()),
+        )),
+    ];
+    let upstream_tcp_address = "localhost:8801".to_string();
+    let upstream_tls_address = "localhost:8802".to_string();
+    setup_bridge_controller(
+        local_server_handle.address(),
+        upstream_tls_address.clone(),
+        subs,
+    )
+    .await;
+    let mut local_client = TestClientBuilder::new(local_server_handle.address())
+        .with_client_id(ClientId::IdWithExistingSession("local_client".into()))
+        .build();
+
+    local_client
+        .subscribe("$internal/connectivity", QoS::AtLeastOnce)
+        .await;
+    // wait to receive subscription ack
+    local_client.subscriptions().recv().await;
+
+    assert_matches!(
+        local_client.publications().recv().await,
+        Some(ReceivedPublication{payload, .. }) if payload == Bytes::from("{\"status\":\"Disconnected\"}")
+    );
+
+    let (mut upstream_server_handle, _) = setup_upstream_broker(
+        AllowAll,
+        Some(upstream_tcp_address.clone()),
+        Some(upstream_tls_address.clone()),
+    );
+
+    assert_matches!(
+        local_client.publications().recv().await,
+        Some(ReceivedPublication{payload, .. }) if payload == Bytes::from("{\"status\":\"Connected\"}")
+    );
+
+    upstream_server_handle.shutdown().await;
+    assert_matches!(
+        local_client.publications().recv().await,
+        Some(ReceivedPublication{payload, .. }) if payload == Bytes::from("{\"status\":\"Disconnected\"}")
+    );
+
+    let (_upstream_server_handle, _) = setup_upstream_broker(
+        AllowAll,
+        Some(upstream_tcp_address),
+        Some(upstream_tls_address),
+    );
+    assert_matches!(
+        local_client.publications().recv().await,
+        Some(ReceivedPublication{payload, .. }) if payload == Bytes::from("{\"status\":\"Connected\"}")
+    );
+}
+
+fn setup_brokers<Z, T>(
+    local_authorizer: Z,
+    upstream_authorizer: T,
+) -> (ServerHandle, ServerHandle, BrokerHandle)
+where
+    Z: Authorizer + Send + 'static,
+    T: Authorizer + Send + 'static,
+{
+    let (local_server_handle, _) = setup_local_broker(local_authorizer);
```

Comment:
nit: might be a good practice to define this broker handle and return it. I see we ignore some of the items currently returned now so we could similarly ignore the local broker handle.